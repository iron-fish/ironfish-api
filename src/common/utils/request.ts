/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Request } from 'express';

export function fetchIpAddressFromRequest(request: Request): string {
  const xForwardedFor = request.header('X-Forwarded-For');
  if (xForwardedFor) {
    const addresses = xForwardedFor.split(',');
    if (addresses.length) {
      return addresses[0].trim();
    }
  }

  // In Heroku, `X-Forwarded-For` will always be set. However, this will be used
  // as a fallback in local development.
  return request.ip;
}
