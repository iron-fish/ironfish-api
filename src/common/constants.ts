/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { EventType } from '.prisma/client';

export const DAYS_IN_WEEK = 7;
export const MS_PER_DAY = 1000 * 60 * 60 * 24;

// Pagination limits
export const DEFAULT_LIMIT = 20;
export const MAX_LIMIT = 100;

// Event categories
export const WEEKLY_POINT_LIMITS_BY_EVENT_TYPE: Record<EventType, number> = {
  [EventType.BLOCK_MINED]: 1000,
  [EventType.BUG_CAUGHT]: 1000,
  [EventType.COMMUNITY_CONTRIBUTION]: 2000,
  [EventType.PULL_REQUEST_MERGED]: 5000,
  [EventType.SOCIAL_MEDIA_PROMOTION]: 1000,
};

export const POINTS_PER_CATEGORY: Record<EventType, number> = {
  [EventType.BLOCK_MINED]: 100,
  [EventType.BUG_CAUGHT]: 100,
  [EventType.COMMUNITY_CONTRIBUTION]: 1000,
  [EventType.PULL_REQUEST_MERGED]: 500,
  [EventType.SOCIAL_MEDIA_PROMOTION]: 100,
};
