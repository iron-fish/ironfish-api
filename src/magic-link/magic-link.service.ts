/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Magic, MagicUserMetadata } from '@magic-sdk/admin';
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

@Injectable()
export class MagicLinkService {
  private readonly magic: Magic;

  constructor(private readonly config: ConfigService) {
    this.magic = new Magic(this.config.get('MAGIC_SECRET_KEY'));
  }

  validate(didToken: string): void {
    this.magic.token.validate(didToken);
  }

  async getMetadataByToken(didToken: string): Promise<MagicUserMetadata> {
    return this.magic.users.getMetadataByToken(didToken);
  }
}
