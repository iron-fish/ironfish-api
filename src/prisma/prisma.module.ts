/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Module } from '@nestjs/common';
import { ApiConfigModule } from '../api-config/api-config.module';
import { PrismaService } from './prisma.service';

@Module({
  exports: [PrismaService],
  imports: [ApiConfigModule],
  providers: [PrismaService],
})
export class PrismaModule {}
