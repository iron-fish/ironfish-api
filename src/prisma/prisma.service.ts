/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Injectable, OnModuleDestroy, OnModuleInit } from '@nestjs/common';
import { PrismaClient } from '@prisma/client';
import { ApiConfigService } from '../api-config/api-config.service';

@Injectable()
export class PrismaService
  extends PrismaClient
  implements OnModuleInit, OnModuleDestroy
{
  constructor(readonly config: ApiConfigService) {
    super({
      datasources: {
        db: {
          url: config.dbPoolUrl,
        },
      },
      // uncomment to log queries
      // log: ['query', 'info', 'warn', 'error'],
    });
  }

  async onModuleInit(): Promise<void> {
    await this.$connect();
  }

  async onModuleDestroy(): Promise<void> {
    await this.$disconnect();
  }
}
