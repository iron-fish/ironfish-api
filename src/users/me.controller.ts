/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Controller, Get, UseGuards } from '@nestjs/common';
import { ApiExcludeController } from '@nestjs/swagger';
import { MagicLinkGuard } from '../auth/guards/magic-link.guard';
import { Context } from '../common/decorators/context';
import { MagicLinkContext } from '../common/interfaces/magic-link-context';
import { SerializedUser } from './interfaces/serialized-user';
import { serializedUserFromRecord } from './utils/user-translator';

@ApiExcludeController()
@Controller('me')
export class MeController {
  @Get()
  @UseGuards(MagicLinkGuard)
  me(@Context() { user }: MagicLinkContext): SerializedUser {
    return serializedUserFromRecord(user);
  }
}
