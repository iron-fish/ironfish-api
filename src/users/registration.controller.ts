/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Controller, Get, Param, Res } from '@nestjs/common';
import { ApiExcludeEndpoint } from '@nestjs/swagger';
import { Response } from 'express';
import { ApiConfigService } from '../api-config/api-config.service';
import { UsersService } from './users.service';

@Controller('registration')
export class RegistrationController {
  constructor(
    private readonly config: ApiConfigService,
    private readonly usersService: UsersService,
  ) {}

  @ApiExcludeEndpoint()
  @Get(':token/confirm')
  async confirm(
    @Param('token') token: string,
    @Res() res: Response,
  ): Promise<void> {
    const user = await this.usersService.findByConfirmationToken(token);
    if (!user || user.confirmed_at) {
      const message = 'Invalid confirmation token';
      res.redirect(
        `${this.config.get<string>(
          'INCENTIVIZED_TESTNET_URL',
        )}/login?toast=${Buffer.from(message).toString('base64')}`,
      );
    } else {
      await this.usersService.confirm(user);
      res.redirect(
        `${this.config.get<string>(
          'INCENTIVIZED_TESTNET_URL',
        )}/login?confirmed=true`,
      );
    }
  }
}
