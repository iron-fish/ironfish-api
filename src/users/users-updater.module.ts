/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Module } from '@nestjs/common';
import { BlocksModule } from '../blocks/blocks.module';
import { PrismaModule } from '../prisma/prisma.module';
import { UsersModule } from './users.module';
import { UsersUpdater } from './users-updater';

@Module({
  exports: [UsersUpdater],
  imports: [BlocksModule, PrismaModule, UsersModule],
  providers: [UsersUpdater],
})
export class UsersUpdaterModule {}
