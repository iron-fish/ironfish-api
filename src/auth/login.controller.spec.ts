/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { HttpStatus, INestApplication } from '@nestjs/common';
import faker from 'faker';
import request from 'supertest';
import { ulid } from 'ulid';
import { v4 as uuid } from 'uuid';
import { MagicLinkService } from '../magic-link/magic-link.service';
import { PrismaService } from '../prisma/prisma.service';
import { bootstrapTestApp } from '../test/test-app';
import { UsersService } from '../users/users.service';

describe('LoginController', () => {
  let app: INestApplication;
  let magicLinkService: MagicLinkService;
  let prisma: PrismaService;
  let usersService: UsersService;

  beforeAll(async () => {
    app = await bootstrapTestApp();
    magicLinkService = app.get(MagicLinkService);
    prisma = app.get(PrismaService);
    usersService = app.get(UsersService);
    await app.init();
  });

  afterAll(async () => {
    await app.close();
  });

  describe('POST /login', () => {
    describe('with no authorization header', () => {
      it('throws a 401', async () => {
        const { body } = await request(app.getHttpServer())
          .post('/login')
          .expect(HttpStatus.UNAUTHORIZED);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with an invalid token', () => {
      it('throws a 401', async () => {
        jest.spyOn(magicLinkService, 'validate').mockImplementationOnce(() => {
          throw new Error('Invalid token');
        });
        const { body } = await request(app.getHttpServer())
          .post('/login')
          .set('Authorization', 'invalid-token')
          .expect(HttpStatus.UNAUTHORIZED);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with no email for the token', () => {
      it('throws a 401', async () => {
        jest
          .spyOn(magicLinkService, 'validate')
          .mockImplementationOnce(jest.fn());
        jest
          .spyOn(magicLinkService, 'getMetadataByHeader')
          .mockImplementationOnce(async () =>
            Promise.resolve({
              issuer: null,
              email: null,
              publicAddress: null,
            }),
          );

        const { body } = await request(app.getHttpServer())
          .post('/login')
          .set('Authorization', 'valid-token')
          .expect(HttpStatus.UNAUTHORIZED);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with a valid token', () => {
      it('updates the last login for a user', async () => {
        const user = await prisma.user.create({
          data: {
            confirmation_token: ulid(),
            email: faker.internet.email(),
            graffiti: uuid(),
            country_code: faker.address.countryCode('alpha-3'),
            confirmed_at: new Date().toISOString(),
          },
        });

        const updateLastLoginAt = jest.spyOn(usersService, 'updateLastLoginAt');
        jest
          .spyOn(magicLinkService, 'validate')
          .mockImplementationOnce(jest.fn());
        jest
          .spyOn(magicLinkService, 'getMetadataByHeader')
          .mockImplementationOnce(async () =>
            Promise.resolve({
              issuer: null,
              email: user.email,
              publicAddress: null,
            }),
          );

        await request(app.getHttpServer())
          .post('/login')
          .set('Authorization', 'valid-token')
          .expect(HttpStatus.OK);

        expect(updateLastLoginAt).toHaveBeenCalledTimes(1);
      });
    });
  });
});
