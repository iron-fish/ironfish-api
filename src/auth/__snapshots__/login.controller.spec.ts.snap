// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginController POST /login with an invalid token throws a 401 1`] = `
Object {
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`LoginController POST /login with no authorization header throws a 401 1`] = `
Object {
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`LoginController POST /login with no email for the token throws a 401 1`] = `
Object {
  "message": "Unauthorized",
  "statusCode": 401,
}
`;
